src/cudamatrix/cu-matrix.cc

```diff
@@ -2540,33 +2577,54 @@ void CuMatrixBase<Real>::Lookup(const std::vector<Int32Pair> &indices,
   // Checks the pointer.
   KALDI_ASSERT(output != NULL);
 
-  // Resizes the output vector.
-  output->resize(indices.size());
+#if HAVE_CUDA == 1
+  CuArray<Int32Pair> cuda_indices(indices);
+  Lookup(cuda_indices, output);
+#else
+  for (int32 i = 0; i < indices.size(); i++) {
+    output[i] = (*this)(indices[i].first, indices[i].second);
+  }
+#endif
+}
+
+template<typename Real>
+void CuMatrixBase<Real>::Lookup(const CuArray<Int32Pair> &indices,
+				Real *output) const {
+  int32 num_elements = indices.Dim();
+  // Checks the dimension.
```

Comment:
Please remove the checks that involve copying the whole vector to CPU.  On the GPU we typically just don't check things like this.